keps/582-preempt-based-on-flavor-order/README.md

```diff
@@ -0,0 +1,330 @@
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Cluster Queue API
+
+We extend the Cluster Queue API to introduce the new fields: flavorFungibility to opt-in and configure the new behavior.
+
+```
+type FlavorFungibility string
+
+const (
+	BeforeNextFlavor FlavorFungibility = "BeforeNextFlavor"
+	AfterAllFlavors  FlavorFungibility = "AfterAllFlavors"
+)
+
+// ClusterQueuePreemption contains policies to preempt Workloads from this
+// ClusterQueue or the ClusterQueue's cohort.
+type ClusterQueuePreemption struct {
+	...
+
+	// +kubebuilder:default="AfterAllFlavors"
+	// +kubebuilder:validation:Enum=BeforeNextFlavor;AfterAllFlavors
+	FlavorFungibility FlavorFungibility `json:"flavorFungibility"`
+}
+```
+
+### Schduling Process in Kueue
+
+For each type of resource in each podSet, Kueue will traverse all resource groups and resource flavors to find a available flavor in present. And this will be the behavior for `AfterAllFlavors`.
+
+If flavorFungibility is BeforeNextFlavor, we will try to allocate podSet's resources in the first flavor in the resource group. Then we will try to preempt in these flavors if some resources not fit. If there are still some resource not fit, we will try the next flavor in the resource group until all resource types get allocated or there is no more resource flavors. 
+
+For example, if cluster queue has 2 resource groups and workload has 1 podSet as the following:
+
+```
+...
+  - coveredResources: ["cpu", "memory"]
+    flavors:
+    - name: "default-flavor1"
+      resources:
+      - name: "cpu"
+        nominalQuota: 3 
+      - name: "memory"
+        nominalQuota: 600Mi 
+    - name: "default-flavor2"
+      resources:
+      - name: "cpu"
+        nominalQuota: 3 
+      - name: "memory"
+        nominalQuota: 600Mi 
+  - coveredResources: ["gpu"]
+    flavors:
+    - name: "vendor1"
+      resources:
+      - name: "gpu"
+        nominalQuota: 9  
+    - name: "vendor2"
+      resources:
+      - name: "gpu"
+        nominalQuota: 9  
+---  
+...
+  podSets:
+  - count: 3
+    spec:
+      containers:
+      - ...
+        resources:
+          requests:
+            cpu: "1"
+            memory: 200Mi
+            gpu: 1
+```
+
+We will first try `default-flavor1` for cpu and memory resources. If `default-flavor1` doesn't fit, we try preempt in `default-flavor1`. If preemptor return preempt fail, we try to allocate resource in `default-flavor2`.
```

Comment:
> If `default-flavor1` doesn't fit, we try preempt in `default-flavor1`. 

In this situation, what happens if we can preempt `default-flavor` but it only frees up resources for 2 pods?

Do we preempt `default-flavor` only If we can free up for all pods? Or, do we preempt `default-flavor` even if we can free up for part of the pods?

Reply:
currently we only allow one kind of resource use one flavor (because now there is only one flavor assignment for one kind of resource).
so I think in this pr we only preempt default-flavor only If we can free up resources for all pods.

Reply:
> because now there is only one flavor assignment for one kind of resource

You're right. We probably should take this situation as another feature.

@alculquicondor @kerthcet Do you think creating an issue for this feature (AssignMultipleFlavorsToWorkload?) is worth it?

Reply:
the preemption logic should not change.
With #806, we can check whether preemption could find enough candidates.
If it's not possible, then we go to the next flavor.

> Do you think creating an issue for this feature (AssignMultipleFlavorsToWorkload?) is worth it?

I don't think we should do this. In most scenarios, this could cause pods to land in very different places, with potentially higher network latency. Also, we can't enforce it, as we wouldn't be able to set node selectors for a subset of pods.

Reply:
> I don't think we should do this. In most scenarios, this could cause pods to land in very different places, with potentially higher network latency. Also, we can't enforce it, as we wouldn't be able to set node selectors for a subset of pods.

That makes sense.